### PR TITLE
Another attempt at comment motion extension

### PR DESCRIPTION
**Pull request by dhleong**
@vlasovskikh Thanks for your work on #105. Here's another try at an extension for providing motion-based comment actions, based off the latest master. This PR will initially serve as more discussion, however, as trying to use the `<Plug>` support is causing some issues that I haven't had time to fully investigate. Basically, the PR mostly works as-is, but without `<Plug>` maps.  When I uncomment the `<Plug>` maps, it seems to want an extra operator before performing the action. 

More concretely, when using the `<Plug>` map indirection: 
- `gcc` works as expected, but it seems to also add an extra `c` as input, such that then hitting `j` will delete the current and next lines and enter insert mode)
- `gcl` to do a block comment around one letter will appear to only accept the `l` to move the cursor. A subsequent `l` will do the comment (but of course at this point it will be one character to the right of the one you actually wanted to comment). In fact, at this point you can enter any motion and it'll "work."
- `gcip` will enter insert mode and paste `g@p`

When directly mapping the keys to the extension handlers:
- The `gc` + motion in normal word works as expected
- `gc` in visual mode doesn't seem to do anything
- `gcc` behaves the same as when using `<Plug>`

The tests, however, all pass. These were observed by using the `runIdea` gradle command from within IntelliJ. The `vim-surround` support seems to work just fine, though, so it could be some problem in my code rather than the extension framework....

Wanted to get this up so we can start looking into the causes of these behaviors, since they'll probably need to be addressed before other extensions can be developed.

### CODE REVIEW

**Comment by vlasovskikh** [comment]
@dhleong Is it an emulation of a known Vim plug-in?

**Comment by domi91c** [comment]
https://github.com/tpope/vim-commentary

**Comment by dhleong** [comment]
It's a subset of tcomment or commentary. I think they both have a lot more
features than the gcc/gc+motion but those are the only ones I'm familiar
with.

**Comment by vlasovskikh** [comment]
@dhleong Thank you. Will you debug the problem with `<Plug>` or do you want me to help with it?

**Comment by dhleong** [comment]
@vlasovskikh I'll see if I have time to investigate this weekend

**Comment by dhleong** [comment]
Got some time today to update some tests so the `gcc` issue is reproduced in a unit. Will try to keep investigating when I can.

**Comment by dhleong** [comment]
Aha. Found the cause of the `gcc` issue:

https://github.com/JetBrains/ideavim/blob/master/src/com/maddyhome/idea/vim/KeyHandler.java#L284

Because there is a previous mapping `gc`, it unnecessarily handles the last-typed `c`. I'm a bit fried right now but sometime later this week I'll see if I can figure out a way to fix this case without breaking the case those lines are originally intended to handle (I see the unit tests).

**Comment by dhleong** [comment]
Just pushed another quick change to the base *AmbiguousMapping tests—they also demonstrate the same issue. Commenting out the if block mentioned above fixes those tests (and this plugin's `gcc` tests), but breaks the other motion-based tests for this plugin.

**Comment by dhleong** [comment]
Okay, fixed the issues with `gcc` with the above push. This includes tests in the `MapCommandTest` as well. The `<Plug>` mapping issues persist, however, as does the issue with `gc` in visual mode. I'll look into these next. If you prefer I make a separate PR to fix the cause of the `gcc` issues, I'd be happy to do that.

**Comment by dhleong** [comment]
Fixed the `<Plug>` issue. Will look into the visual mode stuff later.

**Comment by dhleong** [comment]
Fixed the visual mode `gc` action, with tests. Should be ready for review now! 

Let me know if you want me to pull out the core fixes in a separate PR.

**Comment by dhleong** [comment]
@vlasovskikh Do you need anything else from me on this?

**Comment by vlasovskikh** [comment]
@dhleong I believe I don't. I'll review your PR later, sorry about that.

**Comment by dhleong** [comment]
No worries, just wanted to make sure you weren't blocked on me. Thanks!

**Comment by lpld** [comment]
@vlasovskikh Hello,
Any chances that this will be merged into master?

**Comment by deathbeam** [comment]
Yes please, I miss this feature so much from IdeaVim. Currently, I am using super ugly workaround to at least comment something out:

```viml
nnoremap gcc :action CommentByLineComment<cr>
vnoremap gc :<bs><bs><bs><bs><bs>action VimVisualSwapSelections<cr>:action CommentByLineComment<CR>
```

but having this PR in would make my life a lot easier.

**Comment by ninrod** [comment]
Yes please. I manually comment stuff out with `/* and */ and //`. It feels ultra slow.

**Comment by ayzenquwe** [comment]
Hopefully @vlasovskikh will have some free time to merge it. I also have to use those workarounds mentioned by @deathbeam.

**Comment by jbyler** [comment]
Don't forget one of the benefits of open source: if you can't wait for it to be merged to mainline, you can build the PR branch yourself and install it in your local IntelliJ. For a while I even distributed binaries I built in this way, for the benefit of people who wanted to use features not yet ready for release. You don't even have to write any code--just build it!

**Comment by ninrod** [comment]
hum. @jbyler is it easy to do it? I've never tampered with idea plugins

**Comment by gaganis** [comment]
@ninrod The build process is very easy. You download the source and run the buildPlugin gradle task. This creates the plugin jar in build/libs named ideavim-SNAPSHOT.jar. This can be added to teh IDE from the IntelliJ plugin settings UI using "Install plugin from disk...".

Of cource before building you would need to merge the PR that interests you. I myself usually do it manually which takes a few steps. I think there is support from the Github UI but I have not used it myself (https://help.github.com/articles/merging-a-pull-request/).

If you need help with that please let us know.

**Comment by jbyler** [comment]
For merging the PR, here's a writeup on how I do it, without using a
local branch: http://stackoverflow.com/a/26476083/2562319


Or if you want a local branch to hold the PR code, here's another
way: http://stackoverflow.com/a/18096344/2562319 (also linked from
my answer).

**Comment by ninrod** [comment]
Oh, thanks @jbyler. git was actually the easy part for me. I was afraid of the java/intellij part, but now @gaganis has given me encouragement!

**Comment by vlasovskikh** [comment]
I see a lot of interest in this PR. I'll look into it during my next IdeaVim iteration.

**Comment by ninrod** [comment]
@gaganis, this maybe very stupid, but what is the exact command to run the "buildPlugin gradle task"? thanks in advance.

**Comment by deathbeam** [comment]
@ninrod probably ```./gradlew buildPlugin```

**Comment by vlasovskikh** [comment]
@ninrod `./gradlew buildPlugin`. See `./gradlew help` for more info.

**Comment by gaganis** [comment]
@ninrod The command is what  @deathbeam and @vlasovskikh replied.

Thanks guys for stepping in! :+1:

**Comment by ninrod** [comment]
Thanks guys! This atempts to download a 349MB full version of intellij, but I already have that installed. Is that needed or is there an option to use my installation?

**Comment by ninrod** [comment]
@vlasovskikh, can you share how do you develop ideavim? Do you always build the whole jar, after making a change and the then load it through the settings menu? Or is there a way to run the ideaVim project as a plugin so that the changes are hot-loaded? Is there a way to run ideaVim in debug-mode? Thanks in advance.

**Comment by ayzenquwe** [comment]
@ninrod, everything you need is already in the README.md file, "Where to Start" section. IdeaVim is just a regular IntelliJ IDEA plugin, so you need to refer IDEA's [documentation](http://www.jetbrains.org/intellij/sdk/docs/).

**Comment by aphecetche** [comment]
dear all,

"gc" is also a command I really miss... Any news on this one ?

thanks

**Comment by ninrod** [comment]
@aphecetche, apparently, you have to build it yourself after merging in this branch.

**Comment by mnarrell** [comment]
Any update here?

**Comment by zvirja** [comment]
@dhleong @vlasovskikh Guys, do you know anything about the status of this PR? Comment command is the really missing feature 😟 Probably, somebody else could resume work on this if @dhleong is not interested in it anymore...

**Comment by zvirja** [comment]
@vlasovskikh Yet another try to bring your attention. I know that you are potentially very busy with other stuff, but this feature is really must-have. Other integrations (e.g. VS Code) already have it.

Thanks!

**Comment by vlasovskikh** [comment]
For some reason I keep getting distracted from this PR. I usually track community demands by the amount of votes in the project tracker on YouTrack. There are at least 14 votes here on GitHub. Just in case, I've created the corresponding issue for it in the tracker [VIM-1521](https://youtrack.jetbrains.com/issue/VIM-1521). Feel free to vote for it.

It looks like command-wise this pull request is closer to vim-commentary. There is also a related issue about nerdcommenter-style commands [VIM-922](https://youtrack.jetbrains.com/issue/VIM-922).

**Comment by dhleong** [comment]
I've gone ahead and rebased against master to fix the merge conflicts in anticipation of possibly getting reviewed again

File: src/com/maddyhome/idea/vim/extension/commentary/CommentaryExtension.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        return;
+      }
+
+      // always use line-wise comments
+      if (!new Operator().apply(editor, context, SelectionType.LINE_WISE)) {
+        return;
+      }
+
+      // Leave visual mode
+      executeNormal(parseKeys("<Esc>"), editor);
```
Comment (AlexPl292) [inline src/com/maddyhome/idea/vim/extension/commentary/CommentaryExtension.java]: There is `FLAG_EXIT_VISUAL` for commands. Probably you can use it?

File: src/com/maddyhome/idea/vim/KeyHandler.java [hunk 2]
```diff
@@ -302,8 +302,28 @@ public void run() {
...
+
+          // NB: mappingInfo MUST be non-null here, so if equal
+          //  then prevMappingInfo is also non-null; this also
+          //  means that the prev mapping was a prefix, but the
+          //  next key typed (`key`) was not part of that
+          if (prevMappingInfo == mappingInfo) {
+            // post to end of queue so it's handled AFTER
+            //  an <Plug> mapping is invoked (since that
+            //  will also get posted)
+            Runnable handleRemainingKey = new Runnable() {
```
Comment (AlexPl292) [inline src/com/maddyhome/idea/vim/KeyHandler.java]: Could you replace it with lambda?

File: src/com/maddyhome/idea/vim/extension/commentary/CommentaryExtension.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        PsiFile file = PsiDocumentManager.getInstance(proj).getPsiFile(editor.getDocument());
+        if (file == null) return false;
+
+        handler.invoke(editor.getProject(), editor, editor.getCaretModel().getCurrentCaret(), file);
+        handler.postInvoke();
+
+        // Jump back to start
+        executeNormal(parseKeys("`["), editor);
+        return true;
+      } catch (RuntimeException e) {
```
Comment (AlexPl292) [inline src/com/maddyhome/idea/vim/extension/commentary/CommentaryExtension.java]: Please remove this catch. Occurring runtime exceptions are captured by IntelliJ IDEA and can be sent to JetBrains. But it is not possible if exception is suppressed.

File: test/org/jetbrains/plugins/ideavim/ex/MapCommandTest.java [hunk 4]
```diff
@@ -275,17 +275,29 @@ public void testMapLeader() {
   public void testAmbiguousMapping() {
     configureByText("\n");
     typeText(commandToKeys("nmap ,f iHello<Esc>"));
-    typeText(commandToKeys("nmap ,f2 iBye<Esc>"));
-    typeText(parseKeys(",fh"));
-    myFixture.checkResult("Hello\n");
+    typeText(commandToKeys("nmap ,fc iBye<Esc>"));
```
Comment (AlexPl292) [inline test/org/jetbrains/plugins/ideavim/ex/MapCommandTest.java]: I'm not really understand how this relates to the comment extension.

Comment (dhleong) [reply test/org/jetbrains/plugins/ideavim/ex/MapCommandTest.java]: I honestly don't remember the specifics, but based on the commit messages it looks like I ran into some issues with ambiguous mapping handling—possibly as part of my changes, or maybe just extra cases I wanted to make sure I didn't break.

**Comment by dhleong** [comment]
Thanks for looking at this!

> There are some problems with cursor position after comment. If you disable "Line comment at first column" option for codestyle, cursor position will always be at 0 column. Should it work like this?

No, probably not. I played with my plugin and the cursor seems to stay where it started

> Another problem is that I get `Write access is allowed inside write-action only` exception all the time. Check the red led in the right bottom corner after commenting. 

You're right. The stacktrace seems to be on the `extensionHandler.execute()` line, which I haven't done anything with. The only thing I can think of is that I added another `application.invokeLater`, but it seems to be doing the same thing as the same code below it (`application.invokeLater(handleMappedKeys)`) and I don't get the same error for the Surround extension. I'll have to look into this more; I think the Extension code should ideally be isolated from WriteAction and friends, so there must be something weird I've introduced elsewhere.

> And could you add tests for file type, where it's another comment symbol (not `//`). Perhaps html?

Makes sense.


I went ahead and rebased against master since the `runIdea` gradle bit was broken and I couldn't test properly.

File: src/com/maddyhome/idea/vim/extension/commentary/CommentaryExtension.java (hunk 1 above)
Comment (dhleong) [reply src/com/maddyhome/idea/vim/extension/commentary/CommentaryExtension.java]: It's a `VimExtensionHandler` so I'm not sure where I would set that flag; `VimSurroundExtension` does the same trick.

**Comment by dhleong** [comment]
@AlexPl292 I played with my extension a bit more and it seems that linewise comments keep the cursor in place, while blockwise comments (eg: `/* commented */`) move the cursor to the start of the block comment. I tried to emulate that behavior.

Let me know what else you find!

**Comment by dhleong** [comment]
@AlexPl292 Not to rush you—I understand if you haven't had time to get back here—but are there any requested changes I've missed? This is still labelled as  "Waiting for Updates."

**Comment by AlexPl292** [comment]
Hi @dhleong 
I'm sorry about long reply. I threw myself into work on VIM-510, so I paused PR review for a week or two (I hope this will not take more time). Your PR is the first on my list when I'll continue review.

**Comment by dhleong** [comment]
Hey no worries, thanks for the update! Just wanted to make sure I didn't miss something that was blocking on my end.

**Comment by AlexPl292** [comment]
Okay, I think it's pretty good to merge it into master.
Thank you for PR!

**Comment by AlexPl292** [comment]
I can see the next line in AUTHORS.md: `* [Daniel Leong](mailto:falcone88@gmail.com)`
Is it you? Do you want me to change this email to `me@dhleong.net`?

**Comment by dhleong** [comment]
Yes, that's me. I would appreciate the change—thanks!

**Comment by ninrod** [comment]
Hi, any plans on landing this change in 0.52?

**Comment by AlexPl292** [comment]
@ninrod of course. It's already merged in master.

**Comment by tom-anders** [comment]
Pretty cool, is there a plan to support the dot commant for this?

**Comment by dhleong** [comment]
@tom-anders I proposed an approach with https://github.com/JetBrains/ideavim/pull/110. I'm not sure if it's the best way, and the PR is pretty old, but there is definitely interest in supporting the dot command for this and any other extensions people come up with!

**Pull request merged**